httpclient5/src/test/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperatorTest.java

```diff
@@ -0,0 +1,309 @@
+ *
+ */
+package org.apache.hc.client5.http.impl.nio;
+
+import org.apache.hc.client5.http.DnsResolver;
+import org.apache.hc.client5.http.config.TlsConfig;
+import org.apache.hc.client5.http.nio.AsyncClientConnectionOperator;
+import org.apache.hc.client5.http.nio.ManagedAsyncClientConnection;
+import org.apache.hc.client5.http.ssl.ConscryptClientTlsStrategy;
+import org.apache.hc.core5.concurrent.FutureCallback;
+import org.apache.hc.core5.http.HttpHost;
+import org.apache.hc.core5.http.URIScheme;
+import org.apache.hc.core5.http.config.RegistryBuilder;
+import org.apache.hc.core5.http.nio.ssl.TlsStrategy;
+import org.apache.hc.core5.http.protocol.BasicHttpContext;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.hc.core5.reactor.ConnectionInitiator;
+import org.apache.hc.core5.util.Timeout;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.mockito.stubbing.Answer;
+
+import java.net.InetAddress;
+import java.net.SocketAddress;
+import java.net.UnknownHostException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+
+public class HappyEyeballsV2AsyncClientConnectionOperatorTest {
+
+    private AsyncClientConnectionOperator happyEyeballsV2AsyncClientConnectionOperator;
+
+    private ConnectionInitiator connectionInitiator;
+
+    private DnsResolver dnsResolver;
+
+    private SocketAddress socketAddress;
+
+    private AsyncClientConnectionOperator connectionOperator;
+
+
+    @BeforeEach
+    public void setup() {
+
+        dnsResolver = Mockito.mock(DnsResolver.class);
+        connectionOperator = Mockito.mock(AsyncClientConnectionOperator.class);
+
+        happyEyeballsV2AsyncClientConnectionOperator = new HappyEyeballsV2AsyncClientConnectionOperator
+                (RegistryBuilder.<TlsStrategy>create().register(URIScheme.HTTPS.getId(), ConscryptClientTlsStrategy.getDefault()).build(),
+                        connectionOperator,
+                        dnsResolver,
+                        Timeout.ofSeconds(1),
+                        Timeout.ofSeconds(1),
+                        Timeout.ofSeconds(1),
+                        Timeout.ofSeconds(1),
+                        Timeout.ofSeconds(1),
+                        1,
+                        HappyEyeballsV2AsyncClientConnectionOperator.AddressFamily.IPv4);
+
+        connectionInitiator = mock(ConnectionInitiator.class);
+        socketAddress = mock(SocketAddress.class);
+
+
+    }
+
+    @DisplayName("Test that application prioritizes IPv6 over IPv4 when both are available")
```

Comment:
Again, the sorting rules are more complicated than this. For example, native IPv4 addresses are preferred to IPv6 addresses that use an encapsulation/tunneling transport like 6to4 or Teredo. IPv6 addresses will also be considered unusable, and therefore deprioritized, if no IPv6 _source_ address is available.